src/main/java/org/opensearch/knn/indices/ModelDao.java

```diff
@@ -527,10 +533,56 @@ public void delete(String modelId, ActionListener<DeleteModelResponse> listener)
                 // If model is in Training state, fail delete model request
                 if (ModelState.TRAINING == getModelResponse.getModel().getModelMetadata().getState()) {
                     String errorMessage = String.format("Cannot delete model [%s]. Model is still in training", modelId);
-                    listener.onFailure(new DeleteModelWhenInTrainStateException(errorMessage));
+                    listener.onFailure(new DeleteModelException(errorMessage));
                     return;
                 }
 
+                // Get index mappings to check if model is being used by any indices
+                getMappings(modelId, getMappingsStep);
+            }, listener::onFailure);
+
+            getMappingsStep.whenComplete(getMappingsResponse -> {
+                String modelIdStringInMapping = String.join("model_id=", modelId);
+                Map<String, MappingMetadata> mappings = getMappingsResponse.getMappings();
+                for (Map.Entry<String, MappingMetadata> entry : mappings.entrySet()) {
```

Comment:
Will there be a way to fail fast? like if it is not knn index, skip and move on to the next index?

Reply:
Even on a k-NN index, there isn't necessarily a mapping that indicates if it's a k-NN index.
i.e.
```
{
  "mappings": {
    "properties": {
      "target-field": {
        "type": "knn_vector",
        "model_id": "my-model"
      }
    }
  }
}
```

The fastest fail I could think of to add is a contains() check to see if the model id is present in the string version of the mapping metadata before parsing through the entire map

Reply:
I think we can skip index if index setting "index.knn" is true

Reply:
That would require another transport call to get each index and check the settings.  Is that faster than checking the metadata string?

Reply:
Also, here is an example settings for a target index I created.  I don't see the index.knn setting:
```
{"index.creation_date":"1717708887058","index.number_of_replicas":"1","index.number_of_shards":"1","index.provided_name":"target-index","index.replication.type":"DOCUMENT","index.uuid":"cXM_TDJnSjO62HeJfGWDew","index.version.created":"137217827"}

```